src/include/distributed/listutils.h

```diff
@@ -35,11 +35,38 @@ typedef struct ListCellAndListWrapper
 
 /*
  * foreach_ptr -
- *	  a convenience macro which loops through a pointer list without needing a
- *	  ListCell, just a declared pointer variable to store the pointer of the
- *	  cell in.
+ *	  a convenience macro which loops through a pointer List without needing a
+ *	  ListCell or and index variable, just a declared pointer variable to store
+ *    the iterated values.
  *
- *   How it works:
+ *    PostgreSQL 13 changed the representation of Lists to expansible arrays,
+ *    not chains of cons-cells. This changes the costs for accessing and
+ *    mutating List contents. Therefore different implementations are provided.
+ *
+ *    For more information, see postgres commit with sha
+ *    1cff1b95ab6ddae32faa3efe0d95a820dbfdc164
+ */
+
+#if PG_VERSION_NUM >= PG_VERSION_13
+
+/*
+ *    How it works:
+ *	  - An index is declared with the name {var}PositionDoNotUse and used
+ *	    throughout the for loop using ## to concat.
+ *	  - To assign to var it needs to be done in the condition of the for loop,
+ *	    because we cannot use the initializer since the index variable is
+ *	    declared there.
+ *	  - || true is used to always enter the loop even if var is NULL.
+ */
+#define foreach_ptr(var, l) \
```

Comment:
What do you think of calling adding a macro like foreach_ptr_modify for the dependency.c case where we add to the list while iterating it?

This macro is used all over the code base, so changing it could be a little invasive (e.g. should we re-run performance numbers?).

Hopefully, there are not many places that add to a list that they are iterating.

Reply:
I ran HammerDB benchmark and NOPM decreased to 477458 from 479037 
((479037-477458)/479037) = 0.33% drop in NOPM.

I'll quickly open a seperate PR, and introduce `foreach_ptr_modify`

https://github.com/citusdata/release-test-results/tree/updated_foreach_macros/11_06_2020_1604694466/hammerdb/11_06_2020_1604694466

Reply:
I also ran HammerDB in a smaller cluster with single worker node and NOPM decreased to 8706 from 8907

 (8907-8706) / 8907 = 2.26% drop in NOPM

Reply:
I couldn't really understand how this pr could lead to performance issues :/, can it be a random thing ?